### PR TITLE
W-14646484: prevent error in Min Mule Version calculation

### PR DESCRIPTION
**Pull request by rbourbonnavarro**

### CODE REVIEW

**Comment by rbourbonnavarro** [comment]
--validate

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/parser/java/utils/MinMuleVersionUtils.java [hunk 1]
```diff
@@ -535,6 +560,18 @@ private static ResolvedMinMuleVersion calculateParameterTypeMinMuleVersion(Type
     return parameterMMV;
   }
 
+  private static boolean isLinkageError(Throwable t) {
+    t = t.getCause();
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/parser/java/utils/MinMuleVersionUtils.java:564]: not considering that t itself could be a linkeage errror

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/parser/java/utils/MinMuleVersionUtils.java [hunk 2]
```diff
@@ -535,6 +560,18 @@ private static ResolvedMinMuleVersion calculateParameterTypeMinMuleVersion(Type
...
+    while (t.getCause() != null) {
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/parser/java/utils/MinMuleVersionUtils.java:565]: bad condition. Some times errors end up having themselves as a their cause. If you're going to do this you need a set of seen causes already

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/java/ExtensionWithProvidedDependencyTestCase.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+import org.mule.runtime.api.meta.model.ExtensionModel;
+import org.mule.test.provided.dependency.ProvidedDependencyExtension;
+
+import org.junit.Test;
+
+public class ExtensionWithProvidedDependencyTestCase {
+
+  @Test
+  public void extensionWithParametersRelyingOnProvidedDependencyTypesIsLoaded() {
```
Comment (marianogonzalez) [inline modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/java/ExtensionWithProvidedDependencyTestCase.java:22]: add allure annotations, but most importantly, add a jdoc explaining the details of what you're testing here

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/java/ExtensionWithProvidedDependencyTestCase.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    ExtensionModel extension = loadExtension(ProvidedDependencyExtension.class);
+
+    assertThat(extension.getName(), is(ProvidedDependencyExtension.NAME));
+    assertThat(extension.getMinMuleVersion().isPresent(), is(true));
+    assertThat(extension.getMinMuleVersion().get().toString(), is("4.1.1"));
```
Comment (marianogonzalez) [inline modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/java/ExtensionWithProvidedDependencyTestCase.java:27]: shouldn't you also be asserting that the MMV for the components that couldn't be introspected are correctly set?

**Review by marianogonzalez — CHANGES_REQUESTED** [review]

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/java/ExtensionWithProvidedDependencyTestCase.java (hunk 3 above)
Comment (rbourbonnavarro) [reply modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/java/ExtensionWithProvidedDependencyTestCase.java:36]: Done.

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/parser/java/utils/MinMuleVersionUtils.java (hunk 2 above)
Comment (rbourbonnavarro) [reply modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/parser/java/utils/MinMuleVersionUtils.java]: Done.

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/java/ExtensionWithProvidedDependencyTestCase.java (hunk 4 above)
Comment (rbourbonnavarro) [reply modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/java/ExtensionWithProvidedDependencyTestCase.java:41]: Done.

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/parser/java/utils/MinMuleVersionUtils.java (hunk 1 above)
Comment (rbourbonnavarro) [reply modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/parser/java/utils/MinMuleVersionUtils.java]: Done.

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/parser/java/utils/MinMuleVersionUtils.java [hunk 5]
```diff
@@ -535,6 +560,20 @@ private static ResolvedMinMuleVersion calculateParameterTypeMinMuleVersion(Type
     return parameterMMV;
   }
 
+  private static boolean isLinkageError(Throwable t) {
+    Set<Throwable> throwables = new HashSet<>();
+    while (t != null && !throwables.contains(t)) {
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/parser/java/utils/MinMuleVersionUtils.java:565]: replace with `while (t != null && throwables.add(t))`

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/parser/java/utils/MinMuleVersionUtils.java [hunk 6]
```diff
@@ -535,6 +560,20 @@ private static ResolvedMinMuleVersion calculateParameterTypeMinMuleVersion(Type
...
+      if (t instanceof LinkageError) {
+        return true;
+      }
+
+      throwables.add(t);
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/parser/java/utils/MinMuleVersionUtils.java:570]: remove

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/java/ExtensionWithProvidedDependencyTestCase.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+import io.qameta.allure.Story;
+import org.junit.Test;
+
+@Feature(SDK)
+@Story(MIN_MULE_VERSION)
+@Issue("W-14645134")
+public class ExtensionWithProvidedDependencyTestCase {
+
+  @Test
+  @Description("Tests that the model of an extension that has a custom type that relies on types from a dependency" +
```
Comment (marianogonzalez) [inline modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/java/ExtensionWithProvidedDependencyTestCase.java:33]: add trailing space

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/java/ExtensionWithProvidedDependencyTestCase.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+      "declared as provided, and has that custom type used as a parameter in an operation, is correctly loaded and" +
```
Comment (marianogonzalez) [inline modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/java/ExtensionWithProvidedDependencyTestCase.java:34]: I don't think this quite describes the issue. The provided class was not being used as a parameter but as an internal field. Also, in SDK language, a custom type usually refers to a type that's part of the extensions public API, while in this case it's an internal driver

**Review by marianogonzalez — CHANGES_REQUESTED** [review]

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/parser/java/utils/MinMuleVersionUtils.java (hunk 5 above)
Comment (rbourbonnavarro) [reply modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/parser/java/utils/MinMuleVersionUtils.java:565]: Done.

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/parser/java/utils/MinMuleVersionUtils.java [hunk 9]
```diff
@@ -535,6 +560,20 @@ private static ResolvedMinMuleVersion calculateParameterTypeMinMuleVersion(Type
     return parameterMMV;
   }
 
+  private static boolean isLinkageError(Throwable t) {
+    Set<Throwable> throwables = new HashSet<>();
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/parser/java/utils/MinMuleVersionUtils.java:564]: I would rename to `seen` for clarity

**Review by marianogonzalez — CHANGES_REQUESTED** [review]

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/parser/java/utils/MinMuleVersionUtils.java (hunk 6 above)
Comment (rbourbonnavarro) [reply modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/parser/java/utils/MinMuleVersionUtils.java]: Done.

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/java/ExtensionWithProvidedDependencyTestCase.java (hunk 7 above)
Comment (rbourbonnavarro) [reply modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/java/ExtensionWithProvidedDependencyTestCase.java]: Done.

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/java/ExtensionWithProvidedDependencyTestCase.java (hunk 8 above)
Comment (rbourbonnavarro) [reply modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/java/ExtensionWithProvidedDependencyTestCase.java]: Done.

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/parser/java/utils/MinMuleVersionUtils.java (hunk 6 above)
Comment (rbourbonnavarro) [reply modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/parser/java/utils/MinMuleVersionUtils.java]: Done.

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/parser/java/utils/MinMuleVersionUtils.java (hunk 9 above)
Comment (rbourbonnavarro) [reply modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/parser/java/utils/MinMuleVersionUtils.java]: Done.

**Review by marianogonzalez — DISMISSED** [review]

**Review by marianogonzalez — APPROVED** [review]

**Comment by rbourbonnavarro** [comment]
--validate

**Comment by rbourbonnavarro** [comment]
--validate

**Pull request merged**